src/Internal/DependencyManagement/ExtendedContainer.php

```diff
@@ -43,22 +51,21 @@ class ExtendedContainer extends \League\Container\Container {
 	 * @throws ContainerException Invalid parameters.
 	 */
 	public function add( string $class_name, $concrete = null, bool $shared = null ) : DefinitionInterface {
-		if ( ! $this->class_is_in_root_namespace( $class_name ) && ! in_array( $class_name, $this->registration_whitelist, true ) ) {
-			throw new ContainerException( "Can't use the container to register '$class_name', only objects in the " . Container::WOOCOMMERCE_ROOT_NAMESPACE . ' namespace are allowed for registration.' );
+		if ( ! $this->is_class_allowed( $class_name ) ) {
+			throw new ContainerException( "You cannot add '$class_name', only classes in the {$this->woocommerce_namespace} namespace are allowed." );
 		}
 
-		return parent::add( $class_name, $concrete, $shared );
```

Comment:
Before doing this we need to check that the class either doesn't have a constructor or has one with no arguments, otherwise " Too few arguments to function __construct" will be thrown when trying to resolve the class.

Reply:
I looked at a few options for this:
- `is_callable( array( $class_name, '__construct' ) )` is always `false` due to an idiosyncrasy in PHP.
- `method_exists( $class_name, '__construct' ) )` prevents cases where a constructor may be used to perform initialization such as a `Logger` class that opens a file reference and closes it in `__destruct `.
- We should avoid reflection in the core container wherever possible so we can't check to make sure there's no arguments.

I'm wondering now if my argument against `method_exists()` is invalid given that they can add an inflector with an `init` method. Arguably that may be preferable and it would be easy to add tooling to ensure we never add a constructor to our classes.

Reply:
I think we need to take a step back here.

We have a `set_whatever_dependencies` method that is executed at class creation time, to set the dependencies. And in some cases there might be a need to perform object initialization when it's created, in a `init`-like routine.

The thing is, aren't these two routines basically the same, and by combining them we get a full constructor replacement? So we can just call the routine `init` (or if you want a less generic name, `init_instance` or the like), have it take any required dependencies, and use it for any required initialization too.

And as a bonus, that would settle the discussion on the method name too.

Reply:
Per our conversation we will settle on `init_instance` as a name.

Reply:
I'm documenting here that I noticed some classes use an `init_instance_xxx` and felt a name more descriptive to the specific use-case might be preferable. To that end I've proposed `container_init` since it is an initialization initiated by the container.